examples/ecs/indexing.rs

```diff
@@ -0,0 +1,205 @@
+//! This example demonstrates how to access `Component` data through an `Index`.
+#![allow(clippy::type_complexity)]
+
+use bevy::{
+    ecs::{
+        component::Tick,
+        system::{SystemChangeTick, SystemParam},
+    },
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+use bevy_internal::ecs::query::ReadOnlyWorldQuery;
+use std::{hash::Hash, marker::PhantomData};
+
+pub trait Indexer {
+    type Input: Component;
+    type Index: Hash + Eq + Clone + 'static;
```

Comment:
is this trait bound required? would this not vastly limit the possible Indexers which could be implemented?

seems very HashMap-centric

Reply:
Absolutely want to change these bounds. Right now I'm using `Hash` on the component directly, but I think letting the user provide an `Indexer` trait-object would be ideal, so that they can explicitly decide how the index is generated.

This would also allow the same component to be indexed multiple ways (e.g., a relation pair component could be indexed by each `Entity` value, etc.)